File: app/src/main/java/ui/fragment/CategoriesFragment.java
package ui.fragment;

import android.os.Bundle;
import android.os.Handler;
import android.transition.Slide;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.behavior.HideBottomViewOnScrollBehavior;
import com.google.android.material.button.MaterialButton;

import hlv.cute.todo.R;
import hlv.cute.todo.databinding.FragmentCategoriesBinding;
import ui.adapter.CategoryAdapter;
import ui.dialog.DeleteDialog;
import ui.dialog.MoreDialog;
import utils.DisplayUtils;
import utils.Tags;
import utils.ToastHelper;

public class CategoriesFragment extends BaseFragment {

    private FragmentCategoriesBinding binding;

    private ConstraintLayout toolbar;
    private NestedScrollView nested;
    private RecyclerView rvCategory;
    private FrameLayout frameLytButton;
    private MaterialButton btnAdd;
    private AppCompatImageView imgDeleteAll;
    private HideBottomViewOnScrollBehavior<FrameLayout> scrollBehavior;

    private CategoryAdapter adapter;
    private int scrollYPos = 0;

    public CategoriesFragment() {
    }

    public static CategoriesFragment newInstance() {
        return new CategoriesFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentCategoriesBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
        handleActions();
        handleObserver();
    }

    private void initViews() {
        binding.aImgBack.setOnClickListener(view -> back());
        imgDeleteAll = binding.aImgDeleteAll;
        toolbar = binding.toolbar;
        nested = binding.nested;
        rvCategory = binding.rvCategory;
        frameLytButton = binding.frameLytButton;
        btnAdd = binding.mBtnAdd;

        setScrollBehavior();
        handleShadowScroll();
    }

    private void setScrollBehavior() {
        scrollBehavior = new HideBottomViewOnScrollBehavior<>();
        CoordinatorLayout.LayoutParams lp = ((CoordinatorLayout.LayoutParams) frameLytButton.getLayoutParams());
        lp.setBehavior(scrollBehavior);
    }

    private void handleShadowScroll() {
        nested.setOnScrollChangeListener(new NestedScrollView.OnScrollChangeListener() {
            final float dpShadow = DisplayUtils.getDisplay().dpToPx(nested.getContext(), 12);

            @Override
            public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
                scrollYPos = scrollY;

                if (scrollY == 0) {
                    toolbar.animate().translationZ(0).setStartDelay(0).setDuration(200).start();
                    //toolbar.setTranslationZ(0);
                } else if (scrollY > 50) {
                    toolbar.setTranslationZ(dpShadow);
                    toolbar.animate().translationZ(dpShadow).setStartDelay(0).setDuration(90).start();
                }
            }
        });
    }

    private void handleActions() {
        imgDeleteAll.setOnClickListener(view -> {
            if (getCategoryViewModel().categoriesIsEmpty()) {
                ToastHelper.get().toast(getString(R.string.categories_is_empty));
                return;
            }

            DeleteDialog deleteDialog = new DeleteDialog(getActivity());
            deleteDialog.show();

            deleteDialog.setTitle(getString(R.string.delete_all_categories));
            deleteDialog.setMessage(getString(R.string.delete_all_categories_message, getCategoryViewModel().getCategoriesCount()));
            deleteDialog.setOnClickDelete(() -> {
                getCategoryViewModel().deleteAllCategories();
                scrollBehavior.slideUp(frameLytButton);
                deleteDialog.dismiss();
            });
        });

        btnAdd.setOnClickListener(view -> {
            Fragment fragment = AddEditCategoryFragment.newInstance(null);
            fragment.setEnterTransition(new Slide(Gravity.BOTTOM));
            FragmentTransaction transaction = getParentFragmentManager().beginTransaction();
            transaction.add(R.id.mainContainer, fragment, Tags.FragmentTag.ADD_EDIT_CATEGORY);
            transaction.addToBackStack(Tags.BackStack.ADD_EDIT_CATEGORY);
            transaction.commit();
        });

        handleRecyclerView();
    }

    private void handleRecyclerView() {
        if (getActivity() == null)
            return;

        adapter = new CategoryAdapter(getActivity(),
                category -> {
                    MoreDialog moreDialog = new MoreDialog(getActivity());
                    moreDialog.show();

                    moreDialog.setOnClickDelete(() -> {
                        moreDialog.dismiss();

                        DeleteDialog deleteDialog = new DeleteDialog(getActivity());
                        deleteDialog.show();
                        deleteDialog.setTitle(getString(R.string.delete_category));

                        String categoryName = category.getName();
                        if (categoryName != null && categoryName.trim().length() > 60)
                            categoryName = categoryName.substring(0, 60).trim();

                        deleteDialog.setMessage(getString(R.string.delete_category_message, categoryName));
                        deleteDialog.setOnClickDelete(() -> {
                            getCategoryViewModel().deleteCategory(category);
                            scrollBehavior.slideUp(frameLytButton);
                            deleteDialog.dismiss();
                        });
                    });

                    moreDialog.setOnClickEdit(() -> {
                        moreDialog.dismiss();

                        Fragment fragment = AddEditCategoryFragment.newInstance(category);
                        fragment.setEnterTransition(new Slide(Gravity.BOTTOM));
                        FragmentTransaction transaction = getParentFragmentManager().beginTransaction();
                        transaction.add(R.id.mainContainer, fragment, Tags.FragmentTag.ADD_EDIT_CATEGORY);
                        transaction.addToBackStack(Tags.BackStack.ADD_EDIT_CATEGORY);
                        transaction.commit();
                    });
                }
        );

        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        rvCategory.setLayoutManager(layoutManager);
        rvCategory.setAdapter(adapter);
    }

    private void handleObserver() {
        getCategoryViewModel().fetch();

        getCategoryViewModel().getCategoriesLiveDate().observe(getViewLifecycleOwner(), categories ->
                rvCategory.post(() -> adapter.getDiffer().submitList(categories))
        );

        getCategoryViewModel().getGoToTopLiveData().observe(getViewLifecycleOwner(), scroll ->
                goToTop(1000)
        );
    }

    public void goToTop(int duration) {
        if (nested == null || scrollBehavior == null || frameLytButton == null)
            return;

        nested.smoothScrollTo(0, 0, duration);
        new Handler().postDelayed(() -> scrollBehavior.slideUp(frameLytButton), 500);
    }

    public int getScrollYPos() {
        return scrollYPos;
    }

}

File: app/src/main/java/ui/fragment/CategoriesFragment.kt
package ui.fragment

import android.content.Context
import android.graphics.Point
import android.os.Handler
import android.os.Looper
import android.transition.Slide
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.behavior.HideBottomViewOnScrollBehavior
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.qualifiers.ActivityContext
import hlv.cute.todo.R
import hlv.cute.todo.databinding.FragmentCategoriesBinding
import model.Category
import ui.adapter.CategoryAdapter
import ui.component.PopupMaker
import ui.component.bindingComponent.BaseViewBindingFragment
import ui.dialog.DeleteDialog
import ui.fragment.AddEditCategoryFragment.Companion.newInstance
import utils.Constants
import utils.TextHelper
import utils.ToastUtil
import javax.inject.Inject

@AndroidEntryPoint
class CategoriesFragment : BaseViewBindingFragment<FragmentCategoriesBinding>() {

    override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentCategoriesBinding
        get() = FragmentCategoriesBinding::inflate


    private var scrollBehavior: HideBottomViewOnScrollBehavior<FrameLayout?>? = null

    private var adapter: CategoryAdapter? = null

    var scrollYPos = 0
        private set

    @Inject
    lateinit var toastUtil: ToastUtil

    @Inject
    lateinit var popupMaker: PopupMaker

    @Inject
    @ActivityContext
    lateinit var iContext: Context

    companion object {
        @JvmStatic
        fun newInstance(): CategoriesFragment {
            return CategoriesFragment()
        }
    }

    override fun initiate() {
        initViews()
        handleActions()
        handleObserver()
    }

    private fun initViews() {
        binding.aImgBack.setOnClickListener { back() }

        binding.txtNotesEmpty.text =
            TextHelper.fromHtml(provideResource.getString(R.string.categories_empty_notes))

        setScrollBehavior()
        handleShadowScroll()
    }

    private fun setScrollBehavior() {
        scrollBehavior = HideBottomViewOnScrollBehavior()
        val lp = binding.frameLytButton.layoutParams as CoordinatorLayout.LayoutParams
        lp.behavior = scrollBehavior
    }

    private fun handleShadowScroll() {
        binding.nested.setOnScrollChangeListener(object : NestedScrollView.OnScrollChangeListener {
            val dpShadow = provideResource.getDimen(R.dimen.toolbar_shadow)
            override fun onScrollChange(
                v: NestedScrollView,
                scrollX: Int,
                scrollY: Int,
                oldScrollX: Int,
                oldScrollY: Int
            ) {
                scrollYPos = scrollY
                if (scrollY == 0) {
                    binding.toolbar.animate()
                        .translationZ(0f)
                        .setStartDelay(0)
                        .setDuration(200)
                        .start()

                    //toolbar.setTranslationZ(0);

                } else if (scrollY > 50) {
                    binding.toolbar.translationZ = dpShadow

                    binding.toolbar.animate()
                        .translationZ(dpShadow)
                        .setStartDelay(0)
                        .setDuration(90)
                        .start()
                }
            }
        })
    }

    private fun handleActions() {
        binding.aImgDeleteAll.setOnClickListener {
            if (categoryViewModel.categoriesIsEmpty()) {
                toastUtil.toast(provideResource.getString(R.string.categories_is_empty))
                return@setOnClickListener
            }

            DeleteDialog(iContext).apply {
                show()

                setTitle(provideResource.getString(R.string.delete_all_categories))

                setMessage(
                    provideResource.getString(
                        R.string.delete_all_categories_message,
                        categoryViewModel.getCategoriesCount()
                    )
                )

                onClickDelete = {
                    categoryViewModel.deleteAllCategories()
                    todoViewModel.fetch() //need to update todos if categories was deleted
                    searchViewModel.fetch()
                    scrollBehavior!!.slideUp(binding.frameLytButton)
                    dismiss()
                }
            }
        }

        binding.mBtnAdd.setOnClickListener {
            val fragment: Fragment = newInstance(null).apply {
                enterTransition = Slide(Gravity.BOTTOM)
            }

            parentFragmentManager.beginTransaction().apply {
                add(R.id.mainContainer, fragment, Constants.FragmentTag.ADD_EDIT_CATEGORY)
                addToBackStack(Constants.FragmentTag.ADD_EDIT_CATEGORY)
            }.commit()
        }

        handleRecyclerView()
    }

    private fun handleRecyclerView() {
        adapter = CategoryAdapter(
            iContext,

            onClickMenuListener = { category: Category, anchor: View, wholeItem: View, coordinatePoint: Point? ->

                val popup = popupMaker.showMenu(
                    anchor = anchor,

                    coordinatePoint = coordinatePoint,

                    viewToDim = binding.root,

                    nonDimItem = wholeItem,

                    menuRes = R.menu.popup_menu_categories,

                    onMenuItemClick = itemClicked@{ menuItem ->
                        when (menuItem.itemId) {
                            R.id.menuEdit -> {
                                val fragment: Fragment = newInstance(category).apply {
                                    enterTransition = Slide(Gravity.BOTTOM)
                                }

                                parentFragmentManager.beginTransaction().apply {
                                    add(
                                        R.id.mainContainer,
                                        fragment,
                                        Constants.FragmentTag.ADD_EDIT_CATEGORY
                                    )

                                    addToBackStack(Constants.FragmentTag.ADD_EDIT_CATEGORY)
                                }.commit()
                            }

                            R.id.menuDelete -> {
                                DeleteDialog(iContext).apply {
                                    show()

                                    setTitle(provideResource.getString(R.string.delete_category))

                                    val categoryName = category.name ?: ""

                                    /*if (categoryName != null && categoryName.trim().length > 40)
                                        categoryName = categoryName.substring(0, 40).trim()*/

                                    setMessage(
                                        provideResource.getString(
                                            R.string.delete_category_message,
                                            categoryName
                                        )
                                    )

                                    onClickDelete = {
                                        categoryViewModel.deleteCategory(category)
                                        todoViewModel.fetch() //need to update todos if category was deleted
                                        scrollBehavior!!.slideUp(binding.frameLytButton)
                                        dismiss()
                                    }
                                }
                            }
                        }
                    }
                )

                popupMaker.apply {
                    popup?.changeTextColorOfItem(
                        1,
                        provideResource.getString(R.string.delete),
                        provideResource.getColor(R.color.red)
                    )
                }
            }
        )

        val layoutManager = LinearLayoutManager(iContext)

        binding.rvCategory.apply {
            this.layoutManager = layoutManager
            adapter = this@CategoriesFragment.adapter
        }
    }

    private fun handleObserver() {
        categoryViewModel.fetch()

        categoryViewModel.categoriesLiveDate.observe(viewLifecycleOwner) { categories: List<Category>? ->
            if (categories == null || categories.isEmpty()) {
                binding.rvCategory.visibility = View.GONE
                binding.cLytEmpty.visibility = View.VISIBLE
            } else {
                binding.cLytEmpty.visibility = View.GONE
                binding.rvCategory.visibility = View.VISIBLE
                binding.rvCategory.post { adapter?.differ?.submitList(categories) }
            }
        }

        categoryViewModel.goToTopLiveData.observe(viewLifecycleOwner) {
            goToTop(1000)
        }
    }

    fun goToTop(duration: Int) {
        if (scrollBehavior == null)
            return

        binding.nested.smoothScrollTo(0, 0, duration)

        Handler(Looper.getMainLooper()).postDelayed({
            scrollBehavior!!.slideUp(binding.frameLytButton)
        }, 500)
    }

    override fun onResume() {
        super.onResume()
        categoryViewModel.fetch()
    }
}